edb/pgsql/resolver/command.py

```diff
@@ -339,6 +369,409 @@ def _preprocess_insert_stmt(
+        raise errors.QueryError(
+            'column source is required when inserting into link tables',
+            span=stmt.span,
+        )
+    if not any(c.name == 'target' for c in expected_columns):
+        raise errors.QueryError(
+            'column target is required when inserting into link tables',
+            span=stmt.span,
+        )
+
+    sub_source = link.get_source(ctx.schema)
+    assert sub_source
+    sub_target = link.get_target(ctx.schema)
+    assert sub_target
+
+    # handle DEFAULT and prepare the value relation
+    value_relation, expected_columns = _preprocess_insert_value(
+        stmt.select_stmt, stmt.ctes, expected_columns
+    )
+
+    # if we are sure that we are inserting a single row
+    # we can skip for-loops and iterators, which produces better SQL
+    is_value_single = _has_at_most_one_row(stmt.select_stmt)
+
+    # prepare anchors for inserted value columns
+    value_name = ctx.alias_generator.get('ins_val')
+    iterator_name = ctx.alias_generator.get('ins_iter')
+    source_id = irast.PathId.from_type(
+        ctx.schema,
+        sub_source,
+        typename=sn.QualName('__derived__', value_name),
+        env=None,
+    )
+    link_ref = irtypeutils.ptrref_from_ptrcls(
+        schema=ctx.schema, ptrcls=link, cache=None, typeref_cache=None
+    )
+    value_id: irast.PathId = source_id.extend(ptrref=link_ref)
+
+    value_ql: qlast.PathElement = (
+        qlast.IRAnchor(name=value_name)
+        if is_value_single
+        else qlast.ObjectRef(name=iterator_name)
+    )
+
+    # a phantom relation that is supposed to hold the inserted value
+    # (in the resolver, this will be replaced by the real value relation)
+    value_cte_name = ctx.alias_generator.get('ins_value')
+    value_rel = pgast.Relation(name=value_cte_name)
+    value_columns: List[Tuple[str, bool]] = []
+    for expected_col in expected_columns:
+        if expected_col.name == 'source':
+            ptr_name = 'source'
+            is_link = True
+
+            output_var = pgast.ColumnRef(name=(ptr_name,), nullable=True)
+            value_rel.path_outputs[(source_id, pgce.PathAspect.VALUE)] = (
+                output_var
+            )
+            value_rel.path_outputs[(source_id, pgce.PathAspect.IDENTITY)] = (
+                output_var
+            )
+        elif expected_col.name == 'target':
+            ptr_name = 'target'
+            is_link = True
+
+            ptr_id = value_id.tgt_path()
+            output_var = pgast.ColumnRef(name=(ptr_name,), nullable=True)
+            value_rel.path_outputs[(ptr_id, pgce.PathAspect.VALUE)] = output_var
+            value_rel.path_outputs[(ptr_id, pgce.PathAspect.IDENTITY)] = (
+                output_var
+            )
+        else:
+            # link pointer
+            ptr_name = expected_col.name
+            ptr = link.maybe_get_ptr(ctx.schema, sn.UnqualName(ptr_name))
+            assert ptr
```

Comment:
We want a real error message here, I think? (Unless it is done elsewhere?)

Reply:
No, if this fails, it's a bug.

The ptr_name is coming from a `context.Column`, which means that this column does exist. If this fails, it means that for a given SQL column, we don't know how to map it to an object property.